### PR TITLE
[MXNET-483] C++ tests for mkldnn convolution/deconvolution operator

### PR DESCRIPTION
**Pull request by azai91**
## Description ##
Add tests for MKLDNN convolution / deconvolution operator. Need to convert MKLDNN conv/deconv inputs if they are views (https://github.com/apache/incubator-mxnet/issues/12303).

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Add tests for MKLDNN convolution / deconvolution operator
- [x] Reorder inputs of conv / deconv mkldnn inputs if MKLDNN and view.

## Comments ##
- If this change is a backward incompatible change, why must this change be made.
- Interesting edge cases to note here

### CODE REVIEW

**Comment by azai91** [comment]
@zheng-da can you review

**Comment by azai91** [comment]
@sethman please take a look as well

File: tests/cpp/operator/mkldnn.cc [hunk 1]
```diff
@@ -1204,6 +1271,155 @@ void TestPoolingOp(const OpAttrs &forward_attrs, const OpAttrs &backwards_attrs)
...
+
+      auto tmp_output2 = GetTestInputArrays(true)[i1];
+      NDArray tmp_kernel2 = CreateKernelNDArray(kernel, num_filter, in_arr.arr.shape(), is_deconv);
+      NDArray tmp_bias2 = CreateBiasNDArray(num_filter);
+      backwards_ex_outputs[0] = &tmp_output2.arr;
+      backwards_ex_outputs[1] = &tmp_kernel2;
+      backwards_ex_outputs[2] = &tmp_bias2;
+
+      for (int i = 0; i < backwards_attrs.num_outputs; i++)
+        back_req[0] = kWriteTo;
```
Comment (zheng-da) [inline tests/cpp/operator/mkldnn.cc]: only WriteTo? I think you should test AddTo as well.

Comment (azai91) [reply tests/cpp/operator/mkldnn.cc]: regular fcompute for deconvolution/convolution does not support kAddTo currently
https://github.com/azai91/incubator-mxnet/blob/d79e1ad3294837cac653478045023fd312ceed78/src/operator/nn/convolution-inl.h#L178

https://github.com/azai91/incubator-mxnet/blob/46e47cbc6183d2812a2e405851f0b209383e72ad/src/operator/nn/deconvolution-inl.h#L223

File: tests/cpp/operator/mkldnn.cc [hunk 2]
```diff
@@ -557,33 +614,34 @@ void PrintVerifyMsg(const NDArrayAttrs &arr1, const NDArrayAttrs &arr2) {
...
-  if (dim == 0)
-    slice_amount = num_inputs;
+  int slice_amount = scale[0];
   for (auto shape : shapes) {
-    if (dim >= shape.ndim())
+    if (scale.size() > shape.ndim())
       continue;
-    shape[dim] = shape[dim] * num_inputs;
+
+    for (int dim = 0; dim < scale.size(); dim++)
```
Comment (larroy) [inline tests/cpp/operator/mkldnn.cc]: right type is size_t, ++dim (pre-increment) is standard C++ idiom

File: tests/cpp/operator/mkldnn.cc [hunk 3]
```diff
@@ -557,33 +614,34 @@ void PrintVerifyMsg(const NDArrayAttrs &arr1, const NDArrayAttrs &arr2) {
...
+      shape[dim] = static_cast<int>(round(shape[dim] * scale[dim]));
 
     // Type 1.
     NDArray arr(shape, Context());
     in_arrs.emplace_back(arr, "Normal NDArray");
     InitDefaultArray(&in_arrs.back().arr, rand);
     for (auto pd : pds) {
-      if (num_inputs > 1) {
+      for (int dim = 0; dim < scale.size(); dim++) {
```
Comment (larroy) [inline tests/cpp/operator/mkldnn.cc]: size_t

**Comment by azai91** [comment]
this PR is pending merge of https://github.com/apache/incubator-mxnet/pull/11831

**Comment by azai91** [comment]
kk it was merged. @zheng-da please review

File: src/operator/nn/mkldnn/mkldnn_convolution.cc [hunk 4]
```diff
@@ -290,18 +300,32 @@ void MKLDNNConvolutionBackward(const nnvm::NodeAttrs& attrs, const OpContext &ct
...
+  const NDArray* bias = param.no_bias ? nullptr : &inputs[conv::kBias + 1];
+
+  auto out_grad = inputs[conv::kOut];
+  if (out_grad.IsView() && out_grad.IsMKLDNNData())
+    out_grad = out_grad.Reorder2Default();
+
   mkldnn::convolution_forward::primitive_desc fwd_pd = GetConvFwdImpl(param, ctx.is_train,
-      inputs[conv::kData + 1], inputs[conv::kWeight + 1],
-      param.no_bias ? nullptr : &inputs[conv::kBias + 1], inputs[conv::kOut]);
+      data, weight, param.no_bias ? nullptr : bias, out_grad);
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_convolution.cc]: why do you still need to test param.no_bias?

File: tests/cpp/operator/mkldnn.cc [hunk 5]
```diff
@@ -515,14 +516,81 @@ OpAttrs GetConcatBackwardsOp(int num_args, int dim) {
...
+OpAttrs GetConvBackwardOp(int kernel, int num_filters, int dim, int stride, int pad) {
+  OpAttrs attrs;
+  attrs.attrs.op = Op::Get("_backward_Convolution");
+  attrs.num_inputs = 4;
+  attrs.num_outputs = 3;
+  attrs.attrs.dict.insert({"kernel" , CreateShapeString(kernel, dim)});
+  attrs.attrs.dict.insert({"num_filter" , std::to_string(num_filters)});
+  attrs.attrs.dict.insert({"stride" , CreateShapeString(stride, dim)});
+  attrs.attrs.dict.insert({"pad" , CreateShapeString(pad, dim)});
+  attrs.attrs.op->attr_parser(&attrs.attrs);
```
Comment (zheng-da) [inline tests/cpp/operator/mkldnn.cc]: why no input and output types? same for deconv?

**Comment by zheng-da** [comment]
sorry for the late review.

**Comment by lupesko** [comment]
@azai91 bumping - note the review feedback

**Comment by vandanavk** [comment]
@azai91 ping for addressing review comments.

@sandeep-krishnamurthy Please change label to pr-awaiting-response

File: tests/cpp/operator/mkldnn.cc (hunk 5 above)
Comment (azai91) [reply tests/cpp/operator/mkldnn.cc]: input and output are just specified in the forward operator. the output of the forward is the input of the backwards.

**Comment by ankkhedia** [comment]
@mseth10 @zheng-da ping!
Could you please take a look?

**Comment by lupesko** [comment]
@mseth10 @zheng-da @anirudh2290 - can you guys please take a look? this has been open for a long time. Thanks!

**Comment by mseth10** [comment]
@lupesko Already reviewed and approved.

File: src/operator/nn/mkldnn/mkldnn_deconvolution.cc [hunk 6]
```diff
@@ -197,32 +198,20 @@ MKLDNNDeconvForward::MKLDNNDeconvForward(const DeconvolutionParam& param,
...
+  auto data_mem = in_data.GetMKLDNNDataReorder(
       fwd_pd.diff_dst_primitive_desc());
-  NDArray weight = in_data[deconv::kWeight];
   const mkldnn::memory *weight_mem;
   if (ctx.is_train) {
-    // TODO(zhengda) kvstore doesn't handle MKLDNN correctly. Let's reorder it
-    // to the default format for now.
-    if (weight.IsMKLDNNData())
-      // This asks the engine to reorder data after the weight array is used.
-      weight.Reorder2DefaultAsync();
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_deconvolution.cc]: we want to reorder the layout here. Weights in the kvstore use the default layout.

File: src/operator/nn/mkldnn/mkldnn_deconvolution.cc [hunk 7]
```diff
@@ -197,32 +198,20 @@ MKLDNNDeconvForward::MKLDNNDeconvForward(const DeconvolutionParam& param,
...
     weight_mem = GetWeights(weight, fwd_pd.weights_primitive_desc(), param.num_group);
   } else {
     // For inference, we want to reorder the weight array so we don't need to
     // reorder data every time.
-    if (weight.IsDefaultData()) {
-      weight_mem = GetWeights(weight, fwd_pd.weights_primitive_desc(), param.num_group);
-      // We also need to modify the layout on the original weight array. The
-      // data conversion happens after the weight array is used.
-      weight.MKLDNNDataReorderAsync(fwd_pd.weights_primitive_desc());
```
Comment (zheng-da) [inline src/operator/nn/mkldnn/mkldnn_deconvolution.cc]: In the inference mode, we want only to convert the data layout once to reduce the layout conversion overhead.

**Comment by TaoLv** [comment]
@lupesko I guess @azai91 has merged his changes to #13084 and there is no need to review this one.

**Comment by eric-haibin-lin** [comment]
@TaoLv this PR already got many reviews and one approval. Is it okay to merge this PR, or you have further unaddressed comments?

**Comment by TaoLv** [comment]
@eric-haibin-lin, need @azai91 's inputs about the difference of this PR and #13084.

**Comment by azai91** [comment]
#13084 is a branch off of this PR that includes the BN unit tests. we should merge this one first and then we can more easily review #13084.

File: src/operator/nn/mkldnn/mkldnn_deconvolution.cc (hunk 7 above)
Comment (TaoLv) [reply src/operator/nn/mkldnn/mkldnn_deconvolution.cc]: Same question.

File: tests/cpp/include/test_mkldnn.h [hunk 8]
```diff
@@ -250,6 +254,38 @@ enum ArrayTypes {
   All = 8191,
 };
 
+
+inline NDArray CreateKernelNDArray(TShape kernel, int num_filters, TShape input,
+    bool is_deconv = false) {
+  CHECK(kernel.ndim() == 2) << "mkldnn only supports 2d filters on 4d inputs";
```
Comment (TaoLv) [inline tests/cpp/include/test_mkldnn.h]: CHECK_EQ

**Comment by azai91** [comment]
@TaoLv I checked that this PR does not cause regression.

**Comment by azai91** [comment]
what MKL-DNN warnings? are they introduced in this PR or were they always present? if they were already present, then let's file an issue and tackle in a separate PR.

**Comment by TaoLv** [comment]
Thanks, @azai91 .  I noticed that there were many warnings when built mkldnn cpp test with Makefile. It would be great if you can take a look and fix them. Not necessarily in this PR.

File: src/operator/nn/mkldnn/mkldnn_deconvolution.cc (hunk 6 above)
Comment (TaoLv) [reply src/operator/nn/mkldnn/mkldnn_deconvolution.cc]: Why this was removed?

File: src/operator/nn/mkldnn/mkldnn_deconvolution.cc (hunk 7 above)
Comment (TaoLv) [reply src/operator/nn/mkldnn/mkldnn_deconvolution.cc]: Why this was removed?

**Comment by anirudh2290** [comment]
The warnings are related to batch norm test and implicit_gemm in mshadow. Since they are not being modified in this PR, I think its fine if they are not fixed in this PR.

**Comment by pengzhao-intel** [comment]
@azai91 @anirudh2290 thanks for the help and great works to improve the quality of MKL-DNN.

**Pull request merged**